feat: make analysis tree view more "tree like"

tom-anders opened the pull request:
Especially for heavily commented PGNs this is now much more readable. We're also more consistent with the website now. I added screenhots for comparison below.

I know this is a rather big refactoring, and the recursive nature of this stuff is quite complex, so let me know if anything is unclear. 

I also made a few style adaptions to be a bit more consistent with the the Web UI, but I don't have any strong opinions here, feel free to tweak the styling again after merging.

(Left: Before, Right: This PR)

<p float="left">
  <img src="https://github.com/user-attachments/assets/e21cddc0-8800-4cae-9fcf-7ff53d8cb1ff" width="350" />
  <img src="https://github.com/user-attachments/assets/1e8ca7c3-70cb-4dfa-a64c-d018dfad9dab" width="350" /> 
</p>

lichess.org for comparison:

![Screenshot_2024-09-28_11-23-42](https://github.com/user-attachments/assets/961991a0-5f7e-4690-8f2c-4ce3434f3e16)

tom-anders commented:
(The main trick here that improves the wrapping of comments etc. is to use `Text.rich` to build the lines, instead of the `Wrap` widget)

ijm8710 commented:
This is very cool and agree it looks much better Tom! Thank you! Would you be interested in looking into the two column vertical notation at some point as well? I know it's a different topic but while you're testing notation edge cases, you'll prob be the best man for the job 😀

veloce commented:
Well thanks for this! I know it is tricky to deal with.

Quick review for now: I like it a lot, especially the way you handle sidelines and nesting. It it much more clear now.

Making it look like the web UI, is a non goal to me, as I remind often. The app should have is own visual identity. Of course when the design is better in web we should copy it, as you did here. But we can also vary on some point where it makes sense.

One thing I'm not sure of, for instance, are the italics. Looks like the rule in lichess website is to have an inside sideline italic. Why not... But I think I actually prefer having the comments in italic and keep the inline sideline as is (with a smaller font and a slight opacity). 

I'll review this carefully after I merge the challenges branch (which should be soon now).

tom-anders commented:
> Well thanks for this! I know it is tricky to deal with.

Yeah I stopped counting the knots this put into my brain :D

> Making it look like the web UI, is a non goal to me, as I remind often. The app should have is own visual identity. Of course when the design is better in web we should copy it, as you did here. But we can also vary on some point where it makes sense.
> One thing I'm not sure of, for instance, are the italics. Looks like the rule in lichess website is to have an inside sideline italic. Why not... But I think I actually prefer having the comments in italic and keep the inline sideline as is (with a smaller font and a slight opacity).

Yeah I get a that, I can change it. Using the Web UI as a reference, also for font styles etc. was just more convenient while developing/testing, but yeah no strong opinions on that, I'll adjust the style (or maybe it's even easier if you do it after merging the PR)

veloce commented:
This screen also supports PGN import from random sources. The tree view should look good also in that case.

Could you post a screenshot of the analysis tree when importing this PGN @tom-anders ? https://github.com/lichess-org/dartchess/blob/65fad86b26618785287152e437a7c80efb94b058/data/wcc_2023.pgn#L1-L39

Thanks!

tom-anders commented:
> This screen also supports PGN import from random sources. The tree view should look good also in that case.
> 
> Could you post a screenshot of the analysis tree when importing this PGN @tom-anders ? https://github.com/lichess-org/dartchess/blob/65fad86b26618785287152e437a7c80efb94b058/data/wcc_2023.pgn#L1-L39
> 
> Thanks!

Huh, fun fact (or not fun), I tried importing them into a study via the web interface for comparison, but parsing failed:

![Screenshot_2024-10-02_21-48-51](https://github.com/user-attachments/assets/0cc5d3e1-8626-4b0b-9a63-18a128f3d2f9)

tom-anders commented:
Thanks for that PGN, it uncovered a small bug when building sidelines fixed now. I attached a video where I'm scrolling through the first game of the PGN. 

Note that there are some weird looking line break in some comments, but that's not a wrapping issue, those line breaks are actually there in the source PGN you posted. I'd assume that usually a PGN comment will only have intentional line breaks...?

[treeview.webm](https://github.com/user-attachments/assets/e9279a1d-3971-4190-a7f7-f433ebafa86a)

veloce commented:
Thanks for the vid!
Not sure why there are line breaks in that PGN, let's ignore them.

I find it way more readable in your version, great job!

There's one difference with what I did, the comments are shown in full length, where before only the first 5 lines where shown, and you'd tap on a comment to see the full version in a bottom sheet. I don't know if that change was intentional, but I actually like it, so let's keep it that way!

veloce commented:
One question, have you put a limit on nesting?

I suppose we should have one. Or at least prevent further indentation when reaching a nesting threshold.

tom-anders commented:
> There's one difference with what I did, the comments are shown in full length, where before only the first 5 lines where shown, and you'd tap on a comment to see the full version in a bottom sheet. I don't know if that change was intentional, but I actually like it, so let's keep it that way!

Yeah that was intentional, especially for heavily commented PGNs and studies I thought that it would break your flow if you had to keep tapping comments to read them. 

> One question, have you put a limit on nesting?
> 
> I suppose we should have one. Or at least prevent further indentation when reaching a nesting threshold.

Ah yes good point, haven't implemented that yet - what would be a good threshold after which we stop increasing indentation? Something like 4 maybe? 

Also, we could by default hide deeply nested variations and display a + Icon instead like the website does, what do you think about that?

veloce commented:
> Ah yes good point, haven't implemented that yet - what would be a good threshold after which we stop increasing indentation? Something like 4 maybe?

No more than 4 seems right indeed

> Also, we could by default hide deeply nested variations and display a + Icon instead like the website does, what do you think about that?

👍

tom-anders commented:
> > Also, we could by default hide deeply nested variations and display a + Icon instead like the website does, what do you think about that?
> 
> 👍

Ok I checked the website, it's actually pretty aggressive when it comes to hiding variations: Only two levels of nesting are displayed by default, anything that's three or more levels is hidden by default.

I guess unless we have a good reason not to, let's do the same for now..?

tom-anders commented:
Okay, added an indentation limit, collapse nesting>2 by default and added the plus-icon to expand variations. IMO this makes the "expand variations" action in the move context menu redundant, so I removed it.

"Hide variations" works a little bit different now, it used to only hide the variation you selected the move from, but now it hides all other sibling variations as well. I think it's more useful and intuitive this way, but curious to know what others think

[treeview.webm](https://github.com/user-attachments/assets/7c4856ce-b214-4ad4-aa0e-264281da3161)

veloce commented:
Great work! Yes, I think it should hide all variations. 

I wonder if the "plus" button is easy to tap? Seems a bit small.

Also can you post again the screen recording with the WCC PGN with this new version? thanks!

tom-anders commented:
Yeah you're right, I made the icon bigger now. New video:

[treeview.webm](https://github.com/user-attachments/assets/7e7a31d6-1880-4c14-b134-83e06408e48e)

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,580 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
   }
+}
+
+// A group of parameters that are passed through various parts of the tree view
```
Even for private members we should use the triple slash comments (with a first short line and empty line and description below). It improves code readability a lot, for instance in vim I press "K" and I see the code doc.

tom-anders replied:
Ah, a fellow vim user I see! 

Yeah that was an oversight, will fix

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,580 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+
+    final visibleSideLines = widget.sideLines.whereNot((node) => node.isHidden);
+
+    _keys = List.generate(
```
Doing side effects in `build` is discouraged. Here it looks it can be done in `initState`.

tom-anders replied:
Ah, I initially had this line in `initState`, but then the indent lines would get messed up when hiding/collapsing variations - The solution is to update `_keys` in `didUpdateWidget` as well

veloce reviewed (commented):
There is a lot of changes and this is complex code. I know this part is not much tested yet, but I think it would be nice to add some tests to `analysis_screen_test` relating to what you changed.

Tests don't have to be extra complex first. Using simple PGN with variations, add some tests that check a few basic things:
- when the variation is indented
- when it is inline
- when it is collapsed

It important to start testing this because if later we have bug fixes to make in this tricky screen we'll be happy to have tests. (I shall add more tests for the rest of the analysis feature myself).

One last thing: this is a performance sensitive block. Have you tested the performance using a profile build on a real device? Using large trees to compare with before?

tom-anders commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,580 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  bool shouldShowAnnotations,
+  bool shouldShowComments,
+  GlobalKey currentMoveKey,
+  AnalysisController Function() notifier,
```
hmm me neither, not sure why I did this, will remove the function

tom-anders commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,580 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+        .toList();
+
+    final padding = widget.nesting < 6 ? 12.0 : 0.0;
+    return RepaintBoundary(
```
Without the `RepaintBoundary`, `paint()` is called repeatedly when scrolling (even though it's not necessary to repaint)

tom-anders replied:
(added a comment)

tom-anders replied:
Or maybe this is not a problem? I thought it's probably better for performance this way

veloce replied:
You mean the `_IndentPainter.paint` specifically right?

From flutter doc, on `RepaintBoundary`:
> This widget creates a separate display list for its child, which can improve performance if the subtree repaints at different times than the surrounding parts of the tree.

I don't know if it applies specifically to the indent line widget here. Looks to me there's no reason the indented line subtree would repaint at different times than the surrounding?

tom-anders replied:
Hmm, I'm not 100% what's going on, but currently neither `_IndentPainter.paint()` nor `_IndentedSidelinesState.build()` are called when scrolling. But when removing `RepaintBoundary`, `paint()` gets called, but `_IndentedSidelinesState.build()` doesn't. 

I think https://stackoverflow.com/a/66430673 explains it? 

```
By default CustomPainter is in the same layer as every other widget on the same screen so it's paint method will get called if any other widget on the same screen repaint
```

veloce replied:
Does the removal of `RepaintBoundary` cause a performance issue? If not let's remove it.

tom-anders replied:
I checked, scrolling performance seems to be fine without it, so I removed it

tom-anders commented:
Thanks for the review!

> There is a lot of changes and this is complex code. I know this part is not much tested yet, but I think it would be nice to add some tests to `analysis_screen_test` relating to what you changed.
> 
> Tests don't have to be extra complex first. Using simple PGN with variations, add some tests that check a few basic things:
> 
>     * when the variation is indented
> 
>     * when it is inline
> 
>     * when it is collapsed
> 
>     * ... etc

Sure, I'll add some more tests :+1: 

> One last thing: this is a performance sensitive block. Have you tested the performance using a profile build on a real device? Using large trees to compare with before?

Will do some before/after tests today, the WCC PGN is probably a good test here again

tom-anders commented:
I'll plan to reuse the new `_PgnTreeView` for the study feature, so maybe it's better to make it public, move it to a new file and test it in isolation? what do you think?

tom-anders commented:
Okay, so I did some profiling, by loading the WCC PGN, selecting the first move (e4) and then pressing the "Next" Button, which should rebuild the entire tree view. In general, for both the `main` branch and this PR I observed slight lag, but note that my device has a pretty slow CPU, so I'm seeing lag in a lot of other apps as well. 

Note that I did **not** enable stockfish for these tests (as that would lead to pretty much constant lag anyway)

On the `main` branch, the frame that corresponds to drawing the tree widget takes about 50ms on the UI thread.

With this PR, it instead takes 100-120ms, followed by the frame that draws the indent lines, taking an additional 50-60ms

So yeah, it's a significant performance hit for very large trees like this.

However, I also tested with https://github.com/lichess-org/dartchess/blob/65fad86b26618785287152e437a7c80efb94b058/data/lichess-bullet-game.pgn, which has mostly inline sidelines (but also two indented sidelines due to their length).

With that PGN, `main` takes 15-16ms to render the tree view, while this PR takes 20-22ms, with the next frame taking 6ms to render the indent guides.

tom-anders commented:
Performance update: I'm working on implementing an optimization that only rebuild the part of the mainline (and it's sidelines) that actually changed. Still need to fix some small bugs, but with that fix in place rendering the tree of the WCC PGN now only takes 15-30ms (and rendering the indents becomes negligible) :tada:

veloce commented:
> I plan to reuse the new `_PgnTreeView` for the study feature, so maybe it's better to make it public, move it to a new file and test it in isolation? what do you think?

Ideally we'd include the debouncing and auto-scroll logic, but the debouncing relies on analysis controller. We'd need to refactor to have a common interface for analysis and study (and broadcast).
I think it's a lot more work to do this refactoring. Also I see that `_PgnTreeView` has`AnalysisController` in its parameter, so it shouldn't be compatible with study yet?

Perhaps the easiest for now is to add the tests in the analysis_screen_test (so depending on the analysis controller), and we'll move the tests later when the common interface is done.

veloce commented:
> Performance update: I've implemented an optimization that only rebuilds the subtree of the mainline that actually changed. With this optimization in place rendering the tree of the WCC PGN now only takes 15ms when switching a move within the same mainline subtree, and 30ms when switching the current move from one mainline subtree to another one (since then we need to redraw both the old and the new subtree in that case). Rendering the indents becomes pretty much negligible now. 🎉
> 
> Pretty impressive that with this optimization performance is even better than with the old code

Amazing! Great idea, I hadn't thought about it.

tom-anders commented:
> > I plan to reuse the new `_PgnTreeView` for the study feature, so maybe it's better to make it public, move it to a new file and test it in isolation? what do you think?
> 
> Ideally we'd include the debouncing and auto-scroll logic, but the debouncing relies on analysis controller. We'd need to refactor to have a common interface for analysis and study (and broadcast). I think it's a lot more work to do this refactoring. Also I see that `_PgnTreeView` has`AnalysisController` in its parameter, so it shouldn't be compatible with study yet?

Oh I forgot that I've already written an interface to make the `_PgnTreeView` independent from `AnalysisController` on my study branch, but not on this branch. For `_PgnTreeView` it's easy, because it only calls notifier methods, but does not have any `ref.watch` in it. But also factoring out the debouncing logic is trickier.

So yeah, let's put it into the `analysis_screen_test.dart` for now and refactor in a follow-up PR.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+});
+
+class _PgnTreeViewState extends State<_PgnTreeView> {
+  List<List<ViewNode>> mainlineParts = [];
```
As I'm reading this for the first time, I'm not sure to understand exactly what `mainlineParts` and `treeParts` are for. The types are not enough to self explain.

Some comments explaining the logic would be welcome where needed imo.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+
+      var path = UciPath.empty;
+
+      treeParts = mainlineParts.mapIndexed(
```
There's a big block of code now in the `setState`; Readability would be improved if the code that generates new `treeParts` was externalised in another function.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  required _LineInfo lineInfo,
+  required UciPath pathToNode,
+  required _PgnTreeViewParams params,
+  GlobalKey? moveKey,
```
Not sure to understand what this `moveKey` is? Is it the current move key? If so it'd be better named `currentMoveKey`.

tom-anders replied:
Well, it's only the `currentMoveKey` if this is the current move :D 

but I could simply make this a bool and let the InlineMove assign the key by itself instead, that would probably be more readable

tom-anders replied:
Oh wait, I get confused as well, yeah this definitely needs a comment. This is `firstMoveKey` if it's the start of the sideline (so that we can draw the indent guides) and null otherwise

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+
+  final UciPath initialPath;
+
+  final GlobalKey firstMoveKey;
```
Why is there a need for a `GlobalKey` for the first move? I don't get it.

tom-anders replied:
This is needed so that in `_IndentedSideLinesState` we can query the starting positions of the sidelines and draw the indent lines correctly.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  ];
+}
+
+class _SideLinePart extends ConsumerWidget {
```
I'd add a doc for this class to improve readability. What is it? a part of a sideline?

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  }
+}
+
+class _MainLinePart extends ConsumerWidget {
```
Class doc here too.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  }
+}
+
+class _SideLines extends StatelessWidget {
```
Class doc here too.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  }
+}
+
+class _IndentedSideLines extends StatefulWidget {
```
Class doc

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,654 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+}
+
+class _IndentedSideLinesState extends State<_IndentedSideLines> {
+  late List<GlobalKey> _keys;
```
In flutter we tend to avoid using `GlobalKey` unless having a good reason. Here I'm a little worried about this list of global keys.

I understand you need them to get the position of the sideline to get the proper offset. I think there is a way (no much harder to code I'd say) to do with without all the global keys. The indent width is a fixed one, so in theory you'd need to do sth like:
```dart
final indentOffset = initialOffset + Offset(nestingLevel * fixedIndentWidth, 0)`;
```
So you'd need only one global key to get the initial offset of the tree view, but not more.

tom-anders replied:
The problem is that this only works if the sidelines do not wrap. As soon as a sidelines wraps, we don't know the vertical positions of all following sidelines.

(Or maybe I did not understand your suggestions correctly)

veloce replied:
I suspected something like that, but I had to be sure to understand so pardon me for the question :)

As long as there's no performance issue, this list of global keys shouldn't cause any problem as the scope is limited.

tom-anders replied:
Yeah no problem, I went back and forth about this many times in my head as well, but in the end the problem always was "what if the line wraps though?"

tom-anders commented:
Sorry for the general lack of comments, with complex pieces of code like this that had many iterations until it ended up in its final state, I often find it hard to determine which parts need comments and which don't, especially for non-public classes (since at that point in your head everything seems obvious). 

Added some comments and also renamed stuff, let me know if anything is unclear still.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,696 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+            containsCurrentMove: containsCurrentMove,
+          );
+        } else {
+          // Avoid expensive rebuilds by caching parts of the tree that did not change across a path change
```
One might get confused here: by "expensive rebuilds" you mean actually the framework's `State.build`, but this comment is inside a method named `_rebuildChangedSubtrees`. I think it is worth the precision in the comment.

veloce commented on test/view/analysis/analysis_screen_test.dart:
```diff
@@ -92,13 +99,160 @@ void main() {
+      testWidgets('displays long sideline on its own line', (tester) async {
+        await buildTree(
+          tester,
+          '1. e4 e5 (1... d5 2. exd5 Qxd5 3. Nc3 Qd8 4. d4 Nf6) Nc3 *',
```
Looks like it's missing the `2. ` before `Nc3`.

tom-anders replied:
ah indeed, but looks like our PGN parsing logic is fine with that? But I changed it now

veloce reviewed (commented):
Much better with the documentation! Now I think I understand the logic :)
And thanks for the tests.

veloce commented on lib/src/view/analysis/tree_view.dart:
```diff
@@ -162,124 +131,696 @@ class _InlineTreeViewState extends ConsumerState<AnalysisTreeView> {
+  /// Caches the result of [_mainlineParts] to avoid recalculating it on every rebuild.
+  List<List<ViewNode>> mainlineParts = [];
+
+  /// Caches the top-level subtrees, where each subtree is a [_MainLinePart] and its sidelines.
```
We could add a precision here: this cache is a `Widget` cache obtained from the framework `build` method. The framework will skip `build` of each subtree where the widget reference is the same, this is the real perf optimisation.

Since the other cache above (`mainlineParts`) is not a widget cache it would be good to add this precision to avoid confusion.
